virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+    if sunshine_fraction.size > 1:
+        sunshine_fraction = data["sunshine_fraction"]
+    else:
+        sunshine_fraction = np.ones(len(elevation), dtype=np.float32)
+        LOGGER.info("The sunshine fraction is set to default = 1")
+
+    tau = calculate_atmospheric_transmissivity(elevation, sunshine_fraction)
+    return (1.0 - albedo_shortwave) * tau * shortwave_in
+
+
+def calculate_longwave_radiation(
+    canopy_temperature: NDArray[np.float32],
+    surface_temperature: NDArray[np.float32],
+    canopy_emissivity: float = CANOPY_EMISSIVITY,
+    soil_emissivity: float = SOIL_EMISSIVITY,
+    STEFAN_BOLTZMANN_CONSTANT: float = STEFAN_BOLTZMANN_CONSTANT,
+    CELSIUS_TO_KELVIN: float = CELSIUS_TO_KELVIN,
+) -> NDArray[np.float32]:
+    """Calculate longwave emission from canopy and forest floor, [J m-2].
+
+    Args:
+        canopy_temperature: canopy temperature of n layers, [C]; the array size is
+            set to max number of layers (n_max) and filled with NaN where n < n_max
+        surface_temperature: surface soil temperature, [C]
+        canopy_emissivity: canopy emissivity, default set in config
+        soil_emissivity: soil emissivity, default set in config
+        STEFAN_BOLTZMANN_CONSTANT: Stefan-Boltzmann constant [W m-2 K-4]
+        CELSIUS_TO_KELVIN: factor to convert temperature in Celsius to absolute
+            temperature in Kelvin
+
+    Returns:
+        longwave radiation from n individual canopy layers and soil, [J m-2]
+    """
+    # longwave emission canopy
+    longwave_canopy = (
+        canopy_emissivity
+        * STEFAN_BOLTZMANN_CONSTANT
+        * (CELSIUS_TO_KELVIN + canopy_temperature) ** 4
+    )
+
+    # longwave emission surface
+    longwave_soil = (
+        soil_emissivity
+        * STEFAN_BOLTZMANN_CONSTANT
+        * (CELSIUS_TO_KELVIN + surface_temperature) ** 4
+    )
+
+    # return array of longwave radiation for all canopy layers and surface
+    return np.append(
+        longwave_canopy.transpose(),
+        longwave_soil.reshape([1, len(surface_temperature)]),
+        axis=0,
+    ).transpose()
+
+
+def calculate_netradiation_surface(
+    topofcanopy_radiation: NDArray[np.float32],
+    canopy_absorption: NDArray[np.float32],
+    longwave_radiation: NDArray[np.float32],
+) -> NDArray[np.float32]:
+    """Calculate net shortwave radiation at the surface, [J m-2].
+
+    Args:
+        topofcanopy_radiation: top of canopy radiation shortwave radiation, [J m-2]
+        canopy_absorption: shortwave radiation absorbed by canopy layers, [J m-2]
+        longwave_radiation: longwave radiation from canopy layers and soil, [J m-2]
+
+    Returns:
+        net shortwave radiation at the surface ( = forest floor), [J m-2]
+    """
+    return (
+        topofcanopy_radiation
+        - np.sum(canopy_absorption, axis=1)  # sum over all canopy layers
+        - np.sum(longwave_radiation, axis=1)  # sum over all canopy layers and topsoil
```

Comment:
It's not clear in my head what shapes these various arrays can be. Are they all either vectors or scalars? If so, you shouldn't need the `axis=1` argument, which only makes sense if there are two or more dimensions.

It would be helpful if the docstring mentioned how many dimensions the arguments can/must have.

Reply:
This sparked a discussion around dimension names, the use of xarray, and handling 3D data (time dimension not included). What I initially wanted was to combine the values for the topsoil [1,cell_id] with the canopy values [canopy_layers,cell_id]. To avoid mixup with the vertical axis, we could 1) have all vertical elements in one array (soil +canopy) and select chunks of it if they need to be treated differently, or 2) keep them separate and have two vertical axis